### ISSUE TITLE
Create initial submission for this SDK

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [nimbella/nimbella-sdk-swift#1]
Supporting actions written in Swift is "nice to have" when the front end will also be written in Swift (e.g. for iOS).  I decided to go that route with an app I'm developing.   So, I will (in due course) submit PRs to populate this repo with a functioning SDK for Swift actions under Nimbella.

I note that this repo (created by @rabbah responsive to a suggestion from me) lacks contribution guidelines and perhaps other material needed to provide a legal context for any PR that I might submit.  Rodric, you might want to address this before I do the first PR.   I do have a fork with some skeletal starting points (doesn't do anything useful yet).  That could be a first PR or we could wait for more functionality.

I know that my app will need redis and so I will first add support for that.  This should be easy because there seems to be a good redis client for Swift.   I don't think Nimbella really abstracts redis clients, so the problem just comes down to wrapping the initialization of the client handle with the right auth logic for Nimbella.   After that the client handle is what it is and may or may not align well with redis clients we have for other languages.   This could be improved in the future.   So this will be the first useful PR.

For completeness, I would want to add object store support also.   Adopting the Nimbella storage abstraction for Swift should be quite easy since there is a resemblance between TypeScript and Swift that I can exploit, doing a direct translation of the nodejs SDK's storage support.   Filling in the implementation for S3 should also be easy because AWS provides a Swift client.    That will be the second useful PR.

Google does not provide a google cloud storage client but there do seem to be some authentication clients in Swift that will handle the oauth flows (the hardest part) after which I can use https level support to post the actual requests.   This is more work and I may punt on it or, if not, it will be the final PR.

The app itself will be open-source but limited use (I will probably use ad hoc distribution for it rather than publishing in the Apple app store).   It's purpose is (narrowly) to support scheduling during an annual event I participate in in which large numbers of chamber music players meet in musically-determined groups to play music repertoire.   I think it could also be used during chamber music conferences that involve some "free play", such as the one I attend every summer.    I might try to interest in Associated Chamber Music Players (ACMP) in supporting it, in which case it might end up in the app store after all.  But, usage of the app requires that some app administrator open a Nimbella account to run the backend part.   All of the organizations and usages I have mentioned are non-profit.  I don't know how well the Nimbella free tier will support this sort of usage, but probably the usage will be small and sporadic.

**Comment by rabbah** [comment]
Thanks @joshuaauerbachwatson for the interest. I have added a contribution guide - the same one we use on the other SDKs. The outlined plan of implementation is reasonable. Assuming you will need a Nimbella account hosted in AWS at some point, you can let us know when you're ready and we can enable that for you if it's necessary.

**Comment by joshuaauerbachwatson** [comment]
In https://github.com/nimbella/nimbella-sdk-nodejs/issues/8 I raised some general questions about tests in Nimbella SDKs.   In that issue I noted that redis testing in particular cannot use the redis instances provided by Nimbella directly from clients and so one either needs to install actions to do the testing or install a local redis instance.    Installing actions to do the testing was my original preference, but this is really better reserved for integration testing, because the Swift package manager and the requirements of Nimbella remote build combine to make the incorporation of a new SDK into the action under test awkward (not impossible, just requiring some code-copying preceding the action deployment to be done by a script wrapper, such as a `bats` script).  So, I do plan to supply redis functionality unit tests requiring a local redis instance (based on those in the `redistack` client).

At this point, I have a skeletal implementation in https://github.com/joshuaauerbachwatson/nimbella-sdk-swift but as it does nothing useful I am inclined to hold off on a first PR until I have added redis functionality (including tests).   The skeletal implementation includes a skeletal test (just to get my feet wet with the XCTest framework).   The test indeed verifies that the code does nothing useful 😄 .

**Comment by joshuaauerbachwatson** [comment]
The skeletal implementation in https://github.com/joshuaauerbachwatson/nimbella-sdk-swift now has `redis` functionality filled in and  passes some limited `redis` unit tests using a local redis instance.

I thought the next step would be to make an integration test for the SDK that incorporates it in a Nimbella action (a more realistic test).   This would be followed by the first PR.

However, the `redistack` client requires Swift 5.2 so I can't actually run this test with the current Nimbella Swift runtime, which is at 4.2.   The Nimbella fork of this runtime is behind the Apache master and only goes up to 5.1.   I think this runtime needs to be upgraded to the latest available Apache code (5.4) and then a 5.4 runtime needs to be deployed for Nimbella.

This is going to take a bit of back and forth because
- even though the Nimbella fork of the Swift runtime is public it is unclear whether outside submissions are welcome
- even if I did the work to upgrade this runtime I would not be able to do meaningful tests of the result since I don't have access to Nimbella internal processes
- so, realistically, this upgrade has to be done by someone inside Nimbella

Under the circumstances it might be better to submit a PR with SDK as it is now, even though the "working" parts are only unit tested.   I can then wait for the Swift runtime to be upgraded before verifying that the result really works and asking for it to be incorporated.   Meanwhile, I could possibly make a start on some of the object store parts.

@rabbah please advise on the best path forward.   Depending on weather, I may do some more work on this this week but then I am gone the last two weeks in July for music.   In August, I return to "depending on weather" but without my current hellish practice regime since the music conference will be over.

**Comment by rabbah** [comment]
Contributing incrementally is reasonable given some of the moving pieces that'll be required.

A potential work around for end-to-end testing is to deploy the openwhisk image as a docker action; the image is published here https://hub.docker.com/r/openwhisk/action-swift-v5.4/tags.

**Comment by joshuaauerbachwatson** [comment]
> A potential work around for end-to-end testing is to deploy the openwhisk image as a docker action; the image is published here https://hub.docker.com/r/openwhisk/action-swift-v5.4/tags.

Thanks for the suggestion.  As you may know, I submitted the first PR without doing this.   It seemed to me that this strategy was only incrementally less artificial since there is still the problem of reaching the actual `redis` instances in the NImbella stack and working with Nimbella's convoker conventions for setting up secure `redis` creds.   So I will just wait on the internal work to upgrade the runtime.

**Comment by joshuaauerbachwatson** [comment]
> The app itself will ... support scheduling during an annual event I participate in ...

My app writing plans are out of scope for the issue but I'll continue to record thoughts here (unless told to stop 😄 ).

The scheduling app won't fly for its intended use unless it has a web interface also (a react app?).  This makes the iOS part secondary (and raises the "what about android" question).   I could probably package the web app for android using Cordova (but I could do that for iOS also).   If all of this had been really clear to me a month ago I might not have started on the Swift SDK but, well, I did, and so now it's a fun project in its own right.

To pursue the "all Swift" idea better I'm thinking of first doing a multi-person iPad game.  I actually have one partly developed from my pre-Nimbella days.  It currently uses Apple Mulit-peer Communication (within a household) and I was planning to extend to Apple Game Center.   Testing apps for game center appears difficult (and is a moving target) so I had punted on that part.  I think switching this to serverless would be possible.   I don't know if failing to use game center where it is certainly possible to do so would be grounds for rejection in an app store review.   But, I've gotten relatively little benefit from app store publication so far (all my apps are free anyway), and I'm moving toward using ad-hoc distribution for future apps, so I am not too worried.

This gives rise to the following thought.   When developing a multi-person app for serverless, there will typically be a layer of app-specific id administration.   For example, in the Nimbella stack. the backend would belong to some developer account and one would not want to ask every participant to open a Nimbella account (to give them unique Nimbella auth keys).  So, the developer would probably make the actions public.  This means a secondary layer that is part of the app and which serves to authenticate individual users as part of keeping their state.   Setting this up is a chore and it seems like a potential unit of reuse for future developers.   So, perhaps Nimbella SDKs (eventually all of them) should include a simple user management library with state kept in the developer's key-value and/or object stores.   I don't have any immediate plan to make this part of the Swift SDK effort but will obviously need to do something for any app that I write and we'll see what I come up with.

**Comment by joshuaauerbachwatson** [comment]
It appears that there is now a Swift 5.4 runtime and it is the default for Swift.  However, remote build does not work for it yet because there is no builder action for Swift 5.4.   I've made @rabbah aware of this via Slack and I assume it will get fixed soonish.   I would rely on remote build to build the test actions correctly for doing integration tests of the new SDK.

**Comment by joshuaauerbachwatson** [comment]
It's been a while since I've updated this and meanwhile there is more code in https://github.com/joshuaauerbachwatson/nimbella-sdk-swift/tree/dev.    Time to update on a few fronts.   First, some misc developements.

1.   PR #2 was merged already, so there is already a skeleton there.
2. The problems with the 5.4 runtime and remote build proved to be intermittent and have stopped occurring.   This does not give me a warm feeling but @rabbah and I have agreed to move on and await any future occurrences + insight.
3.  The S3 storage provider is "tentatively code complete" (not meaningfully tested).   This is not the subject of a PR yet since I think _some_ testing (at least unit testing) should be in place first.

Remaining tasks
* [x] Add plausible unit testing for the storage component.   S3 should pass, GCS should fail.
    - technically, this depends on https://github.com/nimbella/nimbella-cli/pull/150 being merged and the result showing up in a stable version.   @rabbah has indicated that the merge is waiting on some legal advice.   I can develop the unit tests without this merge by building my own `nim` so this is not a blocker, but the unit test instructions won't work for other outside developers unless they hand-merge the PR and do a similar build.   Even after the merge a build will be needed until the next stable version.
* [ ] Establish integration testing covering all the unit tested functionality.  Probably using `bats`.   There are some non-trivial obstacles here, which I will cover in a follow-up comment.
* [ ] Add support for the GCS storage provider.   I believe this will be a bit harder than S3 was because there is not a complete storage SDK for swift for GCS.   However, I have not seriously tried yet.

**Comment by joshuaauerbachwatson** [comment]
Some problems with (1) setting up integration tests and (2) usability of the final result for NImbella customers.

1.  I _believe_ that the Swift package manifest (`Package.swift`) lacks the nested scopes concept that one finds with `package.json` for `node`.   If true this means that
    - any user whose build includes a `Package.swift` and wants to use the Nimbella SDK must therein redeclare a direct dependency on it.
    - it's possible that the Nimbella SDK will not be available to single-source Swift builds (but I think this can be addressed by editing the `Package.swift` in the Swift runtime ... that would again require help from @rabbah).
 2.  If the dependency is declared as a source dependency, in a cold container, the time to build the SDK and all its dependencies (which are also source dependencies) will exceed the time limit on builds.   It takes more than 3 minutes on my mac and, empirically, remote builds that do this time out every time.
     - the amount of information fed back in a failed remote build is inadequate, so I am not sure about this.  I am tempted to open an issue on remote build debugging, but I know this problem is basically of my own making ... I should have paid more attention to the issue when I was working on remote build as a Nimbella employee.
3.  A consequence of the previous is that the Nimbella SDK must either
    - be a binary dependency (Swift supports this but documentation on it is a bit vague since source dependencies are their preferred route, at least in Apple documentation) or
    - be cached in some way permanently in the Swift runtime container, details TBD.  Note that this will make integration testing of a _new_ version of the SDK pretty complicated.   Binary dependency seems the better way to go, since it should support both production versions and new versions under test.

**Comment by joshuaauerbachwatson** [comment]
It appears that the swift toolset (and XCode) only offers binary dependencies for Apple targets (iOS and mac, not Linux).   I can see why this might be, since the Linux binary world is very heterogeneous and many-layered (OS versions, C library versions, etc).   I see some evidence online that third parties have solutions but the problem is generally felt to be thorny.

I think the following is the only feasible way forward.

1.  Continue to rely on source dependencies, which means that any "truly cold" build (e.g. when testing a new version of the SDK) is going to take more than the current two minutes timeout for builder actions.
2.  Rely on properly priming the runtime image so that built dependencies are in the cache and builds are (in general) not "truly cold" even in a cold container.

In the short run, though, I need builder actions that have timeouts longer than 2 minutes.   Reprising a conversation with @rabbah, I note that `eksitaly` (where he has given me an account) has a _system maximum_ of 10 minutes which is probably adequate.   However, to take advantage of that higher maximum the builder actions must be re-installed with higher limit specifications (they still time out at 2 minutes).   I don't believe I can override the action limit at invocation time.

**Comment by joshuaauerbachwatson** [comment]
As I do have some unit tests now for storage functionality, and they pass for s3 buckets, I am going to proceed to a PR, since we agreed to roll out the changes gradually.

However, there is _still_ no meaningful integration test and I doubt whether the SDK is usable in the actual Swift runtime yet.   I have opened related issues and have fixes in the pipeline for various problems that keep errors in Swift builds from being seen, but I still do not have the root cause of the actual problem that I am seeing trying to incorporate the SDK into a remotely built Swift action.     The current symptom is that the Swift compiler appears to stop and return exit code 1 after doing all the checkouts and preparation and compiling the first four files.   Yes, it is always four, and not always the same four.    Here are the outputs of several recent runs.

```
... (about 80 seconds in)
Cloning https://github.com/apple/swift-crypto.git
Resolving https://github.com/apple/swift-crypto.git at 1.1.6
Cloning https://github.com/apple/swift-nio-extras.git
Resolving https://github.com/apple/swift-nio-extras.git at 1.10.0
[1/900] Compiling cbs.c
[2/900] Compiling a_int.c
[3/900] Compiling a_i2d_fp.c
[4/900] Compiling a_gentm.c
(stops with exit code 1, no message)
```
```
... (about 80 seconds in)
Cloning https://github.com/apple/swift-nio-transport-services.git
Resolving https://github.com/apple/swift-nio-transport-services.git at 1.11.2
Cloning https://github.com/soto-project/soto.git
Resolving https://github.com/soto-project/soto.git at 5.7.0
[1/857] Compiling a_print.c
[2/857] Compiling a_int.c
[3/857] Compiling a_octet.c
[4/857] Compiling a_i2d_fp.c
(stops with exit code 1, no message)
```
```
... (about 80 seconds in)
Cloning https://github.com/swift-server/async-http-client.git
Resolving https://github.com/swift-server/async-http-client.git at 1.5.0
Cloning https://github.com/apple/swift-nio-extras.git
Resolving https://github.com/apple/swift-nio-extras.git at 1.10.0
[1/857] Compiling a_d2i_fp.c
[2/857] Compiling a_bool.c
[3/857] Compiling a_bitstr.c
[4/857] Compiling a_octet.c
(stops with exit code 1, no message)
```
etc.   Details vary, but it is always stopping after four compiles.  The same build runs locally on my mac with no errors.

I'm fairly stumped at this point.   Even once all the ancillary issues are resolved, allowing build errors to be seen without the need for all sorts of extra steps, I still won't know what's really wrong.

**Comment by joshuaauerbachwatson** [comment]
In an attempt to get to the bottom of the build problem, I ran the gradle build in the dev branch of `nimbella-corp/openwhisk-runtime-swift` to create a local runtime container.  I then used that container to replay a failed remote build of the SDK.  The results were a bit alarming as detailed here:

(1)  The gradle build, when run out of the box, created a clearly incorrect runtime with no copy of `nim` in it.  The gradle build "succeeded" as far as summary messages were concerned.    But, in the transcript was
```
#7 [stage-2  3/16] RUN curl https://apigcp.nimbella.io/downloads/nim/nim-install-linux.sh | bash
#7 sha256:86844ca5f3e02a2c6b5e22858cca68af0e5b5019c8cb04d6eb21281c2cbdb260
#7 0.423 /bin/sh: 1: curl: not found
#7 DONE 0.4s
```
This was easily fixed by tweaking the Dockerfile so as to install `curl` but it gives rise to the question of how the runtime that is presently on `eksitaly` was actually built and how it might differ from the one I built locally.

(2) Once I had a correct (albeit perhaps subtly different) runtime, adjusting it with my `nim` credentials and the right copy of `nim` (with latest mods, merged but not yet in a public stable version) was non-trivial.   I'll deal with that in some other issue as it is part of the general problem of debugging remote builds and not specific to this issue.

(3) Re-running a failed build slice in this container, I did _not_ observe the failure that occurs when running in the actual runtime container.   The build ran to completion and produced a binary artifact!   So, I have to assume the failure in the remote build was environmental, perhaps due to exhausting some resource, and will be very hard to reproduce locally.

(4) Notwithstanding the previous, I did not get a successful deploy.   The binary artifact was apparently too large and I got

```
PUT https://eksitaly.nimbella.io/api/v1/namespaces/jauerbach/actions/test-redis-lite/test?overwrite=true Returned HTTP 413 (Payload Too Large)
```

**Comment by joshuaauerbachwatson** [comment]
It turns out the previous was not totally valid because I mis-identified the artifact to be uploaded, so it was uploading the entire `release` directory, zipped, instead of `release/Action` (the binary).

I first tested the fixed version of the build without using `--remote-build` at all (so it built on my mac).  The result would not have been invocable due to architecture mismatch, BUT it did deploy!    The size of the binary was about 34mb.

I then retested using the slice rebuild approach in my swift runtime container.   It was now trying to upload the right artifact, but it still failed with the size too large error.    The size was about 58mb.   From this I conclude that the limit is somewhere between those numbers but beyond that I don't know how to determine it.  

I don't know why the binary is that much bigger in linux compared to macos but I'm not sure that knowing the answer to that will lead to any solution.

I might be able to get a valid `redis`-only version to build by omitting the s3 storage provider, meaning that the entire `soto` dependency and its transitive closure would be trimmed.  That would be (1) instructive (2) useful to some people who only need `redis`.   I think my app may be ok with just `redis` and no object store.

**Comment by joshuaauerbachwatson** [comment]
> I might be able to get a valid `redis`-only version to build by omitting the s3 storage provider, meaning that the entire `soto` dependency and its transitive closure would be trimmed. That would be (1) instructive

Yes.  This version is on branch `experimental` in my fork.   It is not part of the PR at this point.   It _was_ instructive in that it showed that the redis client is only a small contributor to the overall size.   The size dropped from 58mb to about 10mb and the action deployed.   Not only that: the resulting action was invocable and the invocation returned "success".   Since the action was the integration test for redis, that shows that the redis parts of the SDK actually work.

Note that this was all done by replaying the build in my own local container.   The build _still fails_ for mysterious reasons in the actual runtime running under an invoker.

> (2) useful to some people who only need `redis`

There will be a decision to make on whether to 
1.  drop the object store support entirely and "permanently" so that I can just advertise a useful redis-only SDK 
2. use a conditional build so that I can provide a redis-only version while continuing to figure out how to shrink the `soto` dependency and hopefully eventually adding that support
3. just slog ahead trying to trim the soto dependency before falling back to (1) or (2).

For the moment I am doing 3.

**Comment by joshuaauerbachwatson** [comment]
One bit of low-hanging fruit was the dependency on soto-iam.   I had added this at some point, along with soto-s3, thinking I would need the iam client eventually.   But, I don't believe I do.   The basic client authentication is rolled into the base and the SDK does not need to perform any iam operations.    Taking out that part of the dependency drops the size from 58mg to just under 50mb but that is still too large.

**Comment by joshuaauerbachwatson** [comment]
This is the app that will be the first substantive user of this SDK (if the SDK ends up working; otherwise, I plan to write the new backend for the app in TypeScript):

https://github.com/joshuaauerbachwatson/anyCards

Note that the code as just committed there was sorta-kinda working with proximity communication, but I have not retested with the latest Swift, XCode, etc.   The code does build.  Adding a Nimbella backend is work that has not yet seriously commenced.   If that code is not Swift it will definitely be in a different repo.   Even for Swift, I would need two distinct XCode projects for front and and back end and so I will use the repo structure that most conveniently supports that (TBD).

**Comment by joshuaauerbachwatson** [comment]
I have updated and renamed PR #3.   <strike>I now consider it to be reviewable and mergeable.</strike>   The `README.md` documents the way things are currently, including a `docker`-based workaround for the remote build problems.

The rest of this comment is historical and obsoleted by further developments.   The PR is currently marked as a draft and is still evolving.

<strike>
I consider the work now to be good enough that I will try writing the backend for my app in Swift and I can stumble along even with things in their current state. 

However, there are some tasks that will make it easier for me to use the SDK and will allow it to be offered by Nimbella as a (redis-only for now) toolkit for Swift.   Many of these tasks either require help from Nimbella personnel or have to be done by Nimbella personnel.   I will enumerate them in the next comment.

There are also tasks that I am definitely deferring for the time being.   These are
1.  Adding a GCS storage provider.
2. Expanding the unit test coverage for storage.
</strike>

**Comment by joshuaauerbachwatson** [comment]
This comment is historical.  The actual list of remaining tasks is evolving and is being maintained separately from this issue.

<strike>Remaining tasks to stabilize existing functionality.   A lot of this requires help from inside Nimbella.  For discussion with @rabbah next week.

* [ ] Finish deploying fixed builder actions.   The fix in  https://github.com/nimbella/nimbella-cli/issues/163 is in `eksitaly` but not `nimgcp`.
   - Need someone in NImbella to do this
* [ ] Need more generous time limits on builder actions.
   - timeout currently set to 2 minutes on both `eksitaly` and `nimgcp`
  - I have deployed my own copy of the swift action with a 10 minute timeout on `eksitaly`.  This unblocks me but won't help others.
  - I can't repeat this on `nimgcp` because it won't accept anything longer than 2 minutes.
   - Need someone in NImbella to do this
* [ ] Build deficiencies specific to the Swift runtime
  * [ ] https://github.com/nimbella-corp/openwhisk-runtime-swift/issues/3
    - Important.  Affects all Swift remote builds.  I will submit a PR to fix this but will then need help getting it deployed. 
  * [ ] https://github.com/nimbella-corp/openwhisk-runtime-swift/issues/2
    - Less critical because it only affects error cases.
    - I am reluctant to do a PR because I really don't know Python.  I will do it to get things rolling but someone might have to redo it.
* [ ] Finish rolling out the merged but unpublished changes to CLI and deployer
   - Deployer 3.0.0 was published prior to the merge of the verbose-build fix
   - The CLI is at stable version 1.17.0 which pre-dates all my recent changes.  A new one should be published after the deployer is republished.
   - All runtimes that incorporate `nim` must be rebuilt and redeployed after the previous has been done.  The verbose build fix won't be effective until that is done.
 * [ ] Diagnose and fix the mysterious abrupt failure of remote builds of the SDK and eliminate the current docker-based workaround that I am using to avoid them.
   - These happen even with the fixed builder action and increased timeout and with Swift runtime issue 3 circumvented.   They are not ordinary timeouts.
   - On the other hand, they are probably resource-exhaustion of some sort, given the absence of any other explanation.
   - I need to do more work myself to figure out exactly where I need help.  I'm sure I will, though, because this problem doesn't reproduce in a local container so we will need to rely on system logs, etc.</strike>

**Comment by joshuaauerbachwatson** [comment]
In trying to diagnose the mysterious remote build failure, I found several new problems.   To start with, the remote build started failing fast and very mysteriously.   Here are all the actual problems that led to that.

1.  To support the workaround build I had added a `project.yml` giving the runtime as `swift:5.4`.   This caused the remote copy of `nim` to reject the runtime, which it believed to be invalid.   I'm not sure exactly which version of `nim` is there [subsequently learned it is 1.16] but it either predates the change to interrogate the controller for runtimes or there is some other glitch there.   Changing the runtime in `project.yml` to `swift:default` (which is just as good as far as the workaround is concerned) fixed that problem.
2.  Although we hope that `nim` never actually barfs but always returns a valid result structure (perhaps with failure entries in it), this is not always true.   <strike>For this particular error, there is an unhandled exception that terminates `nim` abruptly with an empty stdout and information on stderr. </strike>.  See subsequent comment.
3. My recent fixes to the builder action did not cover the case where stdout is empty and stderr is not (since I was optimistically assuming that `nim` would never barf ... hah).   This was causing abrupt exit of the action loop.   In general, all of the possible error cases needed tightening. 
4. Even after fixing all the error paths I was still some cases where errors came back as timeouts.   It turns out that it is very important for the output on fd 3 to end in a newline.  This was happening naturally on the "normal" paths but in the error paths it was necessary to add the newline explicitly.

The net is that
- I will have a new version of the builder action to propose, fixing problems (3) and (4)
- I will have more fixes to the deployer and CLI that should go into a stable version and thence into the runtimes.  These will include a fix to problem (2) as the primary motivator, but also some minor things I have spotted along the way.
- I am back to being able to run the build and observe it failing by exiting abruptly in the middle, as it was doing before.

So I am still stumped by the failure itself.

I no longer believe that it always fails after compiling a particular number of files.   This number has varied.   I think the most likely culprit is running out of disk space and so I will investigate that possibility by observing some runs in the local container and characterizing the disk space usage.   Once I have some number and it looks high enough to be a plausible culprit I will need to consult with @rabbah about how to verify that that is the problem remotely and also what to do about it.

**Comment by joshuaauerbachwatson** [comment]
Historical.  See later comments.
<strike>I didn't learn a lot by running the build in a local container.   I would need to learn more linux utilities to see if I can capture the peak usage of
- file descriptors
- subprocesses
- disk space

All of these are possible culprits.   The first two are formally limited by OW.   I could not find information on the third but it stands to reason that the resource is finite.

I think it might be better to approach this from the other direction by taking the activation id of a failed build and seeing if there is a matching event in the invoker logs indicating that one of these resource limits was hit.   I don't have access to the invoker logs so this would require help from @rabbah.</strike>

**Comment by joshuaauerbachwatson** [comment]
> 2\. Although we hope that `nim` never actually barfs but always returns a valid result structure (perhaps with failure entries in it), this is not always true.   For this particular error, there is an unhandled exception that terminates `nim` abruptly with an empty stdout and information on stderr.

I misunderstood what I was seeing.   There is no unhandled exception and `nim` is behaving as designed.   When an error occurs during project reading, or certain non-recoverable errors occur at any point,  `nim` displays the error message on stderr and exits with code 1.

Normally, this looks nothing like an unhandled exception (there is just an error message) but when `nim` is running under a builder action then `--verbose` is set, causing the error to include a stack trace.   A human looking at console output can tell the difference between a handled error with `--verbose` set versus an unhandled exception.   But, when the text is fed back via the stderr log in an activation record, it is very easy to confuse the two.

Two implications.

1.  Although I earlier said I would have a PR to the deployer to close an exception leak, there is no exception leak so no motivation for another PR for that purpose.
2.  There will need to be yet more adjustments to the builder action to treat this kind of error return from `nim` as a more normal occurrence.   I think this probably means also not running with `--verbose` there by default.
3. Probably, after (2) has been worked through, there will be some final changes to the deployer and/or the CLI to make sure the fed-back information is displayed correctly.

**Comment by joshuaauerbachwatson** [comment]
I think the solution to the size problem is embodied here:

https://theswiftdev.com/building-and-loading-dynamic-libraries-at-runtime-in-swift/

If this solution works, it also avoids the problem with abrupt termination of remote builds.  The bulk of the SDK will be built at the time the Swift runtime is built and will reside as a dynamic library inside the container.   The visible SDK that the user links with will be a very small bit of stub code that uses Linux `dlopen` and `dlsym` to load the real thing.   While there probably _is_ some resource problem that affects very large action builds for the Swift runtime, the typical developer is very unlikely to hit it (in any case, it becomes a separable problem and not a blocker for rolling out the Swift SDK).    The 48K limit on the size of the action code stays in effect but, again, developers won't hit it just because they are using the SDK.   To hit the limit they would have to be doing something else "very large" and we can deal with that down the road when it happens.

The ability to produce a dynamic library has always been there in the Swift package manager.   I overlooked it before because it didn't seem to be a solution to the problem (no matter whether the library packaged as static or dynamic, it ends up having to incorporated statically into any executable that is going to be uploaded, since otherwise the code won't be complete).   The insight from the above article is that you can't rely solely on the package manager for dependency resolution if you want any code's loading to be delayed until runtime.   The solution sketched in the article relies on doing this piece yourself, descending to the OS primitives level.

**Comment by joshuaauerbachwatson** [comment]
I have started modifying the PR (which is changed to a draft until it is stable again).   

The `nimbella-object` library product now works in conjunction with two separate loadable modules that contain the bulk of the two storage providers (the GCS is still just a stub that throws "not implemented" but the s3 one is real and unit tests are again passing after the restructure).

I next need to modify the build for the swift runtime to incorporate the two loadable modules and place them in the chosen default place to be dynamically loaded by `nimbella-object`.    The latter should now be pleasingly small.   This should enable the integration test for storage to pass (slightly modified because I can't run the modified runtime in the actual Nimbella stack).

I did not do a similar split (yet) for `nimbella-key-value` because of a specific problem.  The library returns a `RedisClient` object as defined by `RediStack` so the action must import `RediStack` anyway in order to make calls on the result.   Since `RediStack` accounts for most of the size, nothing will be gained by the split until the interface is changed to no longer require a `RediStack` import.   It is less important to do the split for key-value compared to object-store, since the size is already "acceptable."   But, it is still a good idea to do it because the build, even for key-value only, is long running and resource intensive. 

It is probably goodness, anyway, to add a thin abstraction layer, making our handling of redis more like our present handling of storage.   It would be.a chore to make this "complete" for all redis commands because there are a lot of them.   I plan to directly support the commands corresponding to commands under `nim kv` since that is felt to be a useful subset.   There can then be a method to return the actual `RedisClient` object but upcast to `Any` so that `RediStack` does not have to be present when compiling the static portion of the library.   The author of the action will need to import `RediStack` and accept the longer build time if (but only if) he wants to step outside the "starter set" of commands.   For the starter set, this should not be necessary.

The final step, once all the bulky functionality is in dynamically loaded modules, is to recombine `nimbella-object` and `nimbella-key-value` into a single `nimbella-sdk` product once again, for parity with the other SDKs.

**Comment by joshuaauerbachwatson** [comment]
The work in the previous comment is now done.  There is once again a single `nimbella_sdk` import with both object-store and key-value client-making functions.   The substantial logic is dynamically loadable in three libraries, one for `redis`., one for `s3` object stores, and one (dummy) for `gcs` object stores.

I've updated the README consistent with the work to date and in preparation for a roll-out plan.  However, the roll-out plan is becoming pretty complex and I want to test as much of the integration as I can without the ability to run a Nimbella test configuration.   So, I'm leaving the PR as a draft and will mark the companion PR in the Swift runtime as a draft as well.

**Comment by joshuaauerbachwatson** [comment]
Further testing has confirmed that 
- the new factoring "works" (unit tests run correctly)
- the factoring fixes the size problem (actions, no matter which capabilities they use, can be deployed when built in a local docker container and installed from there)
    - they don't actually execute because the dynamic libraries aren't actually present in the currently deployed runtime
    - however, the error message is exactly what one would expect (the action is sound code, it starts up correctly, it fails when trying to load the dynamic library)
- the factoring does _not_ fix the abrupt termination problem in actual remote builds, probably because ...
- ... the build complexity is not substantially changed.   Apparently, dividing a package into products does not keep all of the package's dependencies from being fetched and built, even though only one product is being used and the others are ignored since they aren't statically incorporated.

To complete the fix, I believe I will have to isolate the client code, designed to be linked with each action, into its own package and keep the dynamic libraries in a separate distinct package.

**Comment by joshuaauerbachwatson** [comment]
The separation into two packages worked fine.   The Swift PM requires that a package to be included as a dependency be at the root of its repo.  Thus,  the SDK inclusion as at the root and the second package for the dynamic libraries (as well as the integration tests, which have miniature packages in them) are in subdirectories.   This permits everything to still be in one repo.

With this change, the builds of actions are adequately simple and don't take an inordinate amount of resource to run.

The bad news is that remote builds in the actual runtimes on either `nimgcp` or `eksitaly` are still not working.

The remainder of this comment turns out to be on the wrong track.
<strike>
The good news is that the more streamlined builds make it easy to see that the problem is not resource exhaustion.   Rather, all the C compiles work fine but there is never any output from a Swift compile.   There should be six `.swift` files compiled from the NIO package and a couple of my own.   The compiler exits unexpected trying to compile one of these.   My guess it is one of the NIO ones, and the next step would be to narrow it down.

Just as a minor note, the Swift version in the actual runtimes is 5.4.2 and the one I'm getting in a local Docker build is 5.4.3.  I don't think that's the problem since I have 5.4.2 on my mac and it works fine for the same builds (admittedly with a different target architecture).    But, there might be SOMETHING that got screwed up in the Swift environment of the actual runtimes.  This _could_ just go away by re-installing from the latest code.   I would hold off on that until we are in a position to merge all the open PRs and adopt the new builder actions.</strike>

**Comment by joshuaauerbachwatson** [comment]
> The bad news is that remote builds in the actual runtimes on either nimgcp or eksitaly are still not working.

Oh well, actually, this turns out to be a stupid error.   A while back I decided maybe the problem was memory and in my remote build actions I changed `project.yml` to increase the memory to the max.   Thing is, I omitted the entry for the one action that actually matters in this work (Swift 5.4) and that one action continued to run at the default memory size.   So I concluded that memory wasn't it.  But, it WAS it, or at least a part of it.   Increasing the memory to 768m enabled the (now pretty streamlined) remote build to run.

Of course, the resulting action won't execute because the new dynamic libraries are not yet built into the deployed runtime.

But, I'm now pretty close to having all the pieces lined up for the final push (consisting of getting PRs reviewed and merged and doing the necessary deployment steps for runtimes, actions, and the CLI).

**Comment by joshuaauerbachwatson** [comment]
Because I know @rabbah and others involved with the Nimbella stack are very busy, I've decided to make PR#3 self-contained, in the sense that
1.  The integration tests will run with only workarounds in the tests themselves, not requiring building artifacts from other repos.
2. The resulting SDK will be actually useful for both redis and S3, using similar localized workarounds.
3. The workarounds can be dropped when related changes in other repos are merged and deployed.   This can be done in a more leisurely way.

To do this, I have added the `ensureLibrary` API, expected to be temporary, which downloads the dynamic library(s) needed for a test.   Unfortunately, this can't be done at build time (since the action will not necessarily run in the same container incarnation as the build).   So, the time to do the download adds to execution time.   Definitely not a permanent solution but will go away when the fixes to the runtime are merged and deployed.

**Comment by joshuaauerbachwatson** [comment]
I hit what is, for me at this time, a brick wall.   I feel foolish that when I added the dynamic library support, I tested only on Mac and not on Linux (which is where it matters).    Having belatedly tested on Linux about a week ago, I saw that I am getting seg-faults on Linux from code that works fine on Mac and I don't know why.   I have spent a week exhausting obvious explanations.

One thing I was not able to do is to use `lldb` to examine the seg-fault context directly.  I suspect that would shed light, although it might not be obvious what one needs to do in Swift code to fix it.   The `lldb` utility is present in the Swift runtime container but apparently the Linux in that container is configured to disallow `ptrace`.   When work is resumed on this, if there is still this problem, we should definitely look at enabling `ptrace` so that low level debugging is possible.

Observations:  

1.  the Swift eco-system is most mature around iOS and Mac and lags on Linux.   Apple's investment in the Linux implementation is close to nothing and it's up to the OSS community to keep it up.   In this case, I suspect that the `Unmanaged` class (which is needed to bridge between the typed and untyped domains) differs on Linux in some way that I _could_ possibly figure out but don't want to at this point.
2. the dynamic loading solution, even on Mac (where it seems to work) is not officially endorsed by Apple anyway.

Implication: 

To have a Swift SDK that is maintainable and won't end up driving DO/Nimbella crazy, we need to stay within a "safe" area of things known to work, even if that means going slower on rolling out functionality.

Plan:

1.  In the related PR nimbella-corp/openwhisk-runtime-swift#5, I will roll it back to omit the step to add dynamic SDK libraries.   The PR also contains other things that should be merged.
2. I will close PR #3 since it is based off my `dev` branch which is now hacked to death.    I want to preserve that branch for a possible resumption of this work in the future.
3. I will start a new branch from commit 6ebd5b4, which is just before I started creating dynamic libraries.   On that branch I will remove the S3 support and create a clean redis-only SDK submission which will become a new PR.   It will use static linking.   Redis does not really need dynamic linking because the static size is only about 10 meg.  Remote builds using the SDK will, yes, take longer, but I may be able to ameliorate that by priming the cache in the runtime image.   I _believe_ I will not have the old remote build problem because that was actually due to not enough memory.  I will verify that before opening the new PR.

**Comment by joshuaauerbachwatson** [comment]
The redis-only version of the PR (#4) has been ready to merge for a while.    Since this has no hard dependency on changes to the Swift runtime, it can be used even without any deployment steps within the Nimbella stack (see instructions in the README).  It does require that you create your own builder actions, because the memory limit on the deployed ones is too small.   Other pending PRs (e.g. to the runtime) are "useful to have" because of improved error handling behavior but are not needed to run correct code.

Meanwhile, I _finally_ determined what is causing the seg-faults on Linux that prevented me from including the s3 storage provider.in the PR.   I only got insight after to building exactly the toy demo described in https://theswiftdev.com/building-and-loading-dynamic-libraries-at-runtime-in-swift/ and determining that it, too, gets the seg-fault when run in Linux.  So, the problem is not in my adaptation of the idea in my own code but rather in the idea itself!   Well, sort of.   Actually, the offending code is the correct-looking and innocuous-looking code that follows, which I thought was a good idea so I did the same thing 😄 .  Who wants resource leaks, after all?

```swift
       defer {
            dlclose(openRes)
        }
```

(strictly speaking, the above code won't compile in Linux because in the `dlclose` signature there won't accept `nil` as a possible value.   You have to ensure that the argument isn't an `optional` so I had to reorganize the demo code slightly even to run the test).

The problem is that the `dlclose` executes when the loading subroutine exits its scope, not when the usage of the resolved symbol goes out of scope.   That's how `defer` works in Swift ... it's dependent on the compile-time scoping.   I only learned that when I was able to turn on `lldb` debugging in the container (this requires some extra flags on `docker run`) and actually single step through the code.   I speculate that `dlclose` in MacOS does not necessarily release the storage of the loaded code but in Linux it does, so you do _not_ want to execute a `dlclose` while still using the loaded code.   Simply removing that clause from the code allows it to work.

I am going to try the same thing with my own code and this may enable me to add the s3 storage provider to the PR.

**Comment by joshuaauerbachwatson** [comment]
The change discussed in the previous comment works for the SDK too.   So, I now have a version (on the `dev` branch of my fork) that includes the s3 storage provider and passes integration tests.   The bad news is that there is a fragility concerning the lifecycle of open storage client handles.   The integration test creates two of them, one for the web bucket and one for the data bucket.   In the original version of the test, the second was assigned to a mutable variable that had contained the first (causing the ref count on the first to drop to zero).    The second handle then behaved in a broken fashion.   I think what is happening is that the Soto AWSClient, when operated the default way, has only one EventLoop instance.   When the first handle de-initializes, it drives cleanup code that terminates that event loop.

The test passes now only because I am being careful to keep both handles live for the duration of the test.

The fragility is unacceptable but I think the solution may be fairly simple.   I either have to cache Soto's AWSClient somewhere so that all storage handles share it or I have to use options in the AWSClient constructor that allow me to control EventLoop instances myself.
